src/tactic/unify_equations.lean

```diff
@@ -0,0 +1,415 @@
+
+lemma add_add_one_ne (n m : ℕ) : n + (m + 1) ≠ n :=
+begin
+  apply ne_of_gt,
+  apply nat.lt_add_of_pos_right,
+  apply nat.pos_of_ne_zero,
+  contradiction
+end
+-- Linarith could prove this, but I want to avoid that dependency.
+
+/--
+`match_n_plus_m n e` matches `e` of the form `nat.succ (... (nat.succ e')...)`.
+It returns `n` plus the number of `succ` constructors and `e'`. The matching is
+performed up to normalisation with transparency `md`.
+-/
+meta def match_n_plus_m (md) : ℕ → expr → tactic (ℕ × expr) :=
+λ n e, do
+  e ← whnf e md,
+  match e with
+  | `(nat.succ %%e) := match_n_plus_m (n + 1) e
+  | _ := pure (n, e)
+  end
+
+/--
+Given `equ : n + m = n` or `equ : n = n + m` with `n` and `m` natural numbers
+and `m` a nonzero literal, this tactic produces a proof of `false`. More
+precisely, the two sides of the equation must be of the form
+`nat.succ (... (nat.succ e)...)` with different numbers of `nat.succ`
+constructors. Matching is performed with transparency `md`.
+-/
+meta def contradict_n_eq_n_plus_m (md : transparency) (equ lhs rhs : expr) :
+  tactic expr := do
+  ⟨lhs_n, lhs_e⟩ ← match_n_plus_m md 0 lhs,
+  ⟨rhs_n, rhs_e⟩ ← match_n_plus_m md 0 rhs,
+  is_def_eq lhs_e rhs_e md <|> fail
+    "internal error in contradict_n_eq_n_plus_m:\nexpected {lhs_e} and {rhs_e} to be definitionally equal at transparency {md}.",
+  let common := lhs_e,
+  guard (lhs_n ≠ rhs_n) <|> fail
+    "internal error in contradict_n_eq_n_plus_m:\nexpected {lhs_n} and {rhs_n} to be different.",
+  -- Ensure that lhs_n is bigger than rhs_n. Swap lhs and rhs if that's not
+  -- already the case.
+  ⟨equ, lhs_n, rhs_n⟩ ←
+    if lhs_n > rhs_n
+      then pure (equ, lhs_n, rhs_n)
+      else do {
+        equ ← to_expr ``(eq.symm %%equ),
+        pure (equ, rhs_n, lhs_n)
+      },
+  let diff := lhs_n - rhs_n,
+  let rhs_n_expr := reflect rhs_n,
+  n ← to_expr ``(%%common + %%rhs_n_expr),
+  let m := reflect (diff - 1),
+  pure `(add_add_one_ne %%n %%m %%equ)
+
+/--
+Given `equ : t = u` with `t, u : I` and `I.sizeof t ≠ I.sizeof u`, we solve the
+goal by contradiction.
+-/
+meta def unify_cyclic : unification_step :=
+λ equ type _ _ _ lhs_whnf rhs_whnf _,
+do {
+  -- Establish `sizeof lhs = sizeof rhs`.
+  sizeof ← get_sizeof type,
+  hyp_lhs ← to_expr ``(%%sizeof %%lhs_whnf),
+  hyp_rhs ← to_expr ``(%%sizeof %%rhs_whnf),
+  hyp_type ← to_expr ``(@eq ℕ %%hyp_lhs %%hyp_rhs),
+  hyp_proof ← to_expr ``(@congr_arg %%type ℕ %%lhs_whnf %%rhs_whnf %%sizeof %%equ),
+  hyp_name ← mk_fresh_name,
+  hyp ← note hyp_name hyp_type hyp_proof,
+
+  -- Derive a contradiction (if indeed `sizeof lhs ≠ sizeof rhs`).
+  falso ← contradict_n_eq_n_plus_m semireducible hyp hyp_lhs hyp_rhs,
+  exfalso,
+  exact falso,
+  pure goal_solved
+} <|>
+pure not_simplified
+
+/--
+Sequence two unification steps. First run step `s`. If this was successful
+(i.e. it simplified or solved the goal), stop and return the result of `s`.
+Otherwise run `t` and return its result.
```

Comment:
I'm not so convinced by the description / name: for sequencing I would expect `t` to run when `s` is successful. Also they are `unification_step`s, not simplifiers. How about something like `orelse_step`?